Skip backing indices with a disjoint range on @timestamp field.

martijnvg opened the pull request:
Implicitly skip backing indices with a time series range that doesn't match with
a required filter on @timestamp field.

Relates to #74660

martijnvg committed:
Skip backing indices with a disjoint range on @timestamp field.

Implicitly skip backing indices with a time series range that doesn't match with
a required filter on @timestamp field.

Relates to #74660

martijnvg committed:
fix mistake

martijnvg committed:
Merge remote-tracking branch 'es/master' into skip_backing_indices

martijnvg committed:
added first test

martijnvg committed:
checkstyle

martijnvg commented:
(this needs more tests)

martijnvg commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContext.java:
```diff
@@ -38,24 +39,26 @@ public CoordinatorRewriteContext(
 
     boolean hasTimestampData() {
-        return indexLongFieldRange.isComplete() && indexLongFieldRange != IndexLongFieldRange.EMPTY;
+        return timeSeriesRange != null || (indexLongFieldRange.isComplete() && indexLongFieldRange != IndexLongFieldRange.EMPTY);
```
👍 Makes sense. I will do that and make this PR production ready.

martijnvg committed:
Merge remote-tracking branch 'es/master' into skip_backing_indices

martijnvg committed:
fix test compile error after merging in master

martijnvg committed:
Merge remote-tracking branch 'es/master' into skip_backing_indices

martijnvg committed:
iter

martijnvg committed:
spotless

martijnvg committed:
iter

martijnvg committed:
Merge remote-tracking branch 'es/master' into skip_backing_indices

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

martijnvg committed:
refactor

martijnvg committed:
fix bug, in case of IndexLongFieldRange.EMPTY the hasTimestampData() method should return false.

martijnvg committed:
added test and
ensure field type of @timestamp of backing indices of tsdb data streams are loaded as well.

martijnvg committed:
RemoveTimeSeriesRange with IndexLongFieldRange

martijnvg committed:
re-order

martijnvg commented on server/src/main/java/org/elasticsearch/indices/TimestampFieldMapperService.java:
```diff
@@ -143,6 +143,12 @@ private static boolean hasUsefulTimestampField(IndexMetadata indexMetadata) {
             return false;
         }
+
+        if (indexMetadata.getTimeSeriesTimestampRange() != null) {
```
After writing more tests and actually debugging the code, I noticed that I needed to add this if statement here. Otherwise `@timestamp` range queries against tsdb indices always execute can_match phase against nodes contain shards instead of resolving wether the query resolves to `MatchNoneQueryBuilder` on coordinating node.

I have no idea how to write a good test for this, that checks that shards are skipped because on the coordinating node we figured out that a query resolves to a `MatchNoneQueryBuilder`. 

The point of this PR, is that we should figure out on coordinating if a query has a range query against `@timestamp` field that doesn't match with the the start_time and end_time of indices. I can't check whether can_match search phase skipped shards or the query rewrite on the coordinating node. I would like to verify this,.

martijnvg replied:
Only `@timestamp` field of indices that are listed in `TimestampFieldMapperService` can do the rewrite on coordinating node if understand the code correctly.

nik9000 replied:
Ah! That makes sense.

This is the kind of thing I'd test with a yaml test if possible. It's sneaky enough that I'd want an "outside" test rather than a unit test. And it's important for performance.

nik9000 replied:
Hmmmm. Adding this means we're going to be capturing the mapper for the timestamp of every tsdb index on every node. I wonder if in a follow up we can make these singletons....

martijnvg replied:
> Hmmmm. Adding this means we're going to be capturing the mapper for the timestamp of every tsdb index on every node. I wonder if in a follow up we can make these singletons....

Yes. I think we can do that if the timestamp field mapping between data streams are configured in the same way. Caching the same timestamp field mapper is silly.

martijnvg replied:
> Is it available in the profiler?

I don't think we record this anywhere in `CanMatchPreFilter`. Would be great if we keep track of can match request like we do for query and fetch request in `SearchStats`. Then we can assert in tests that can_match counts haven't increased after doing a search.

martijnvg committed:
Merge remote-tracking branch 'es/master' into skip_backing_indices

martijnvg committed:
fix test compile error after merging in master

martijnvg committed:
Merge remote-tracking branch 'es/master' into skip_backing_indices

martijnvg committed:
Merge remote-tracking branch 'es/master' into skip_backing_indices

martijnvg committed:
use start and end time from IndexMetadata instead of parsing it from index settings

martijnvg committed:
Add unit like test that verifies that for tsdb data stream
we have an entry in TimestampFieldMapperService, so that
search request with a required range query on `@timestamp`
field can skip indices/shards on coordinating node.

martijnvg committed:
Merge remote-tracking branch 'es/master' into skip_backing_indices

martijnvg committed:
spotless

martijnvg committed:
fix bug

martijnvg committed:
Merge remote-tracking branch 'es/master' into skip_backing_indices

martijnvg committed:
handle when index mode is provided in upper case

martijnvg committed:
Merge remote-tracking branch 'es/master' into skip_backing_indices

martijnvg committed:
spotless

martijnvg commented:
run elasticsearch-ci/part-1

nik9000 commented on server/src/main/java/org/elasticsearch/index/IndexMode.java:
```diff
@@ -185,6 +192,13 @@ public IdFieldMapper buildIdFieldMapper(BooleanSupplier fieldDataEnabled) {
+        public IndexLongFieldRange getConfiguredTimestampRange(IndexMetadata indexMetadata) {
+            long min = indexMetadata.getTimeSeriesStart().toEpochMilli();
+            long max = indexMetadata.getTimeSeriesEnd().toEpochMilli();
+            return IndexLongFieldRange.NO_SHARDS.extendWithShardRange(0, 1, ShardLongFieldRange.of(min, max));
```
Hey. So. Like. I think I made `getTimestampBound` and it does like nearly the same thing, right? I feel like that method should be involved in this PR somehow. Or maybe `IndexSettings.timestampBounds` should be.

martijnvg replied:
Good point. I think I can reuse `getTimestampBound(...)` and then `IndexMetadata#getTimeSeriesTimestampRange(...)` use that to build a `IndexLongFieldRange` instance.

martijnvg replied:
`TimestampBounds` needs `IndexScopedSettings` which I don't have in the context In want to use it.
This technically only needed for dynamically updating endTime field, which is my case isn't needed.
But maybe with a minor change to `TimestampBounds`,  I can use it too.

martijnvg replied:
Yes, I can do that.

martijnvg commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into skip_backing_indices
